src/structs/jpeg_write.rs

```diff
@@ -357,59 +357,73 @@ fn encode_block_seq(
+    // to efficiently skip zero blocks using trailing zero scan.
+    let b: &[i16x16; 4] = cast_ref(block.get_block());
+    let mut mask = (b[0].cmp_eq(i16x16::ZERO).move_mask() as u64)
+        | ((b[1].cmp_eq(i16x16::ZERO).move_mask() as u64) << 16)
+        | ((b[2].cmp_eq(i16x16::ZERO).move_mask() as u64) << 32)
+        | ((b[3].cmp_eq(i16x16::ZERO).move_mask() as u64) << 48);
+
+    // flip the bits since cmp_eq returns 0xffff for zero coefficients
+    mask = !mask;
+
+    // already processed DC coefficient, so skip it
+    mask >>= 1;
     let mut bpos = 1;
-    // encode AC
-    while bpos < 64 {
-        // if nonzero is encountered
-        let mut tmp = block[bpos];
-        bpos += 1;
 
-        if tmp == 0 {
-            let mut z = 1;
+    // encode ACs
+    while mask != 0 {
+        let mut zeros = mask.trailing_zeros();
 
-            loop {
-                if bpos == 64 {
-                    huffw.write(actbl.c_val[0x00].into(), actbl.c_len[0x00].into());
-                    return;
-                }
+        if zeros > 15 {
+            // JPEG encoding only supports 15 zeros in a row. Most implementations
+            // write 0xf0 codes for 16 zeros in a row, but we don't need
+            // a special case since write_coef with a zero coefficient
+            // and a 0xf zero count will write the correct code.
+            zeros = 15;
+        }
 
-                tmp = block[bpos];
-                bpos += 1;
+        bpos += zeros + 1;
+        mask >>= zeros + 1;
 
-                if tmp != 0 {
-                    // if we have 16 or more zero, we need to write them in blocks of 16
-                    while z >= 16 {
-                        huffw.write(actbl.c_val[0xF0].into(), actbl.c_len[0xF0].into());
-                        z -= 16;
-                    }
-                    write_coef(huffw, tmp, z, actbl);
-                    break;
-                }
+        write_coef(
+            huffw,
+            block.get_coefficient((bpos - 1) as usize),
+            zeros,
+            actbl,
+        );
 
-                z += 1;
-            }
-        } else {
-            write_coef(huffw, tmp, 0, actbl);
+        if bpos >= 64 {
+            // if we get all 64 coefficients, we're done and don't need an EOB
+            return;
         }
     }
+
+    // write EOB since we didn't get all 64 coefficients
+    huffw.write(actbl.c_val[0x00].into(), actbl.c_len[0x00].into());
 }
 
 /// encodes a coefficient which is a huffman code specifying the size followed
 /// by the coefficient itself
 #[inline(always)]
-fn write_coef(huffw: &mut BitWriter, coef: i16, z: u8, tbl: &HuffCodes) {
+fn write_coef(huffw: &mut BitWriter, coef: i16, z: u32, tbl: &HuffCodes) {
     // vli encode
     let (n, s) = envli(coef);
-    let hc = ((z & 0xf) << 4) + s;
+
+    // compiler is smart enough to figure out that this will never be >= 256,
+    // so no bounds check
+    let hc = (z << 4 | s) as usize;
```

Comment:
It will be better to reverse `z` and `s` in the index: z is [0,15] while s is [0,11] - you will have better cache utilization.

Reply:
tried this but it didn't make a difference in performance. I guess the lookup tables are small enough anyway.

Reply:
Could you push this variant? I'll check performance on my machine.